File: sentinel/scorer.py
"""Bayesian scam scoring with Thompson Sampling for signal weight learning."""

import json
import logging
import math
import os
import random

from sentinel.models import JobPosting, RiskLevel, ScamSignal, SignalCategory, ValidationResult

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Learned-weight cache (loaded once per process from DB)
# ---------------------------------------------------------------------------

_learned_weights_cache: dict[str, float] | None = None

# Minimum observations before a learned weight overrides the static default.
_MIN_OBSERVATIONS = 10


def _load_learned_weights(db_path: str = "") -> dict[str, float]:
    """Query the DB for all active patterns and return {signal_name: bayesian_weight}.

    The result is cached at module level so the DB is only hit once per
    process.  If the DB is unavailable or empty, returns an empty dict and
    scoring falls back to static weights.
    """
    global _learned_weights_cache
    if _learned_weights_cache is not None:
        return _learned_weights_cache

    weights: dict[str, float] = {}
    try:
        from sentinel.db import SentinelDB  # deferred to keep import lightweight

        db = SentinelDB(path=db_path) if db_path else SentinelDB()
        rows = db.get_patterns(status="active")
        for row in rows:
            obs = row.get("observations", 0)
            if obs < _MIN_OBSERVATIONS:
                continue
            alpha = row.get("alpha", 1.0)
            beta = row.get("beta", 1.0)
            total = alpha + beta
            if total <= 0:
                continue
            # Use pattern_id as the key — it matches signal names in the DB
            pid = row.get("pattern_id", "")
            name = row.get("name", "")
            bayesian_w = alpha / total
            if pid:
                weights[pid] = bayesian_w
            if name and name != pid:
                weights[name] = bayesian_w
        db.close()
    except Exception:
        # DB missing, locked, or schema mismatch — fall back gracefully
        logger.debug("Could not load learned weights from DB; using static defaults", exc_info=True)
        weights = {}

    _learned_weights_cache = weights
    return _learned_weights_cache


def _reset_learned_weights_cache() -> None:
    """Clear the module-level cache (used by tests and after flywheel cycles)."""
    global _learned_weights_cache
    _learned_weights_cache = None


# ---------------------------------------------------------------------------
# Core scoring
# ---------------------------------------------------------------------------

def score_signals(
    signals: list[ScamSignal],
    *,
    db_path: str = "",
    use_learned_weights: bool = True,
) -> tuple[float, float]:
    """Compute scam score and confidence from extracted signals.

    Uses weighted Bayesian combination:
    - Positive signals reduce score
    - Red flags and warnings increase score
    - Confidence is based on number and agreement of signals

    When *use_learned_weights* is True (the default), the scorer will try to
    load Bayesian posterior weights from the DB and use them in place of the
    hard-coded signal weights for any pattern that has >= 10 observations.

    Returns (scam_score 0-1, confidence 0-1)
    """
    if not signals:
        return 0.0, 0.0

    _POSITIVE = SignalCategory.POSITIVE

    # Try loading learned weights from the DB
    learned: dict[str, float] = {}
    if use_learned_weights:
        try:
            learned = _load_learned_weights(db_path=db_path)
        except Exception:
            learned = {}

    # Log-odds accumulation: start at log-odds 0 (= 50% prior)
    log_odds = 0.0
    for s in signals:
        # Check for a learned weight override (match on signal name)
        if s.name in learned:
            w = learned[s.name]
        else:
            w = s.weight

        w = max(1e-6, min(1.0 - 1e-6, w))
        # Positive signals move log-odds toward "legitimate"
        if s.category == _POSITIVE:
            # Positive signal: weight is chance-of-scam, so (1-w)/w is the
            # legitimacy odds ratio — subtract it to push score toward 0.
            log_odds -= math.log((1.0 - w) / w)
        else:
            log_odds += math.log(w / (1.0 - w))

    scam_score = 1.0 / (1.0 + math.exp(-log_odds))

    # Confidence: grows with signal count, penalised when positive and negative
    # signals strongly disagree (mixed evidence lowers confidence).
    n_scam = sum(1 for s in signals if s.category != _POSITIVE)
    n_pos = sum(1 for s in signals if s.category == _POSITIVE)
    total = len(signals)

    # Base confidence from signal count (asymptote ~0.95)
    base_conf = 1.0 - math.exp(-0.3 * total)

    # Agreement factor: 1.0 when all same-direction, lower when mixed
    if total > 0:
        majority = max(n_scam, n_pos)
        agreement = majority / total
    else:
        agreement = 1.0

    confidence = round(base_conf * agreement, 4)
    scam_score = round(scam_score, 4)

    return scam_score, confidence


def classify_risk(scam_score: float) -> RiskLevel:
    """Map score to risk level."""
    if scam_score < 0.2:
        return RiskLevel.SAFE
    if scam_score < 0.4:
        return RiskLevel.LOW
    if scam_score < 0.6:
        return RiskLevel.SUSPICIOUS
    if scam_score < 0.8:
        return RiskLevel.HIGH
    return RiskLevel.SCAM


def build_result(
    job: JobPosting,
    signals: list[ScamSignal],
    analysis_time_ms: float = 0.0,
) -> ValidationResult:
    """Full pipeline: score signals, classify risk, build result."""
    scam_score, confidence = score_signals(signals)
    risk_level = classify_risk(scam_score)
    return ValidationResult(
        job=job,
        signals=signals,
        scam_score=scam_score,
        confidence=confidence,
        risk_level=risk_level,
        analysis_time_ms=analysis_time_ms,
    )


# ---------------------------------------------------------------------------
# Bayesian weight learner
# ---------------------------------------------------------------------------

class SignalWeightTracker:
    """Bayesian weight learner for signal effectiveness.

    Uses Beta(alpha, beta) posteriors per signal name.
    Updated when user reports confirm/deny our predictions.
    Thompson Sampling for exploration.
    """

    def __init__(self) -> None:
        # Maps signal name -> (alpha, beta); Beta(1,1) = uniform prior
        self._weights: dict[str, tuple[float, float]] = {}

    def _get_posterior(self, signal_name: str) -> tuple[float, float]:
        return self._weights.get(signal_name, (1.0, 1.0))

    def update(self, signal_name: str, was_correct: bool) -> None:
        """Update posterior: alpha += was_correct, beta += (not was_correct)."""
        alpha, beta = self._get_posterior(signal_name)
        if was_correct:
            alpha += 1.0
        else:
            beta += 1.0
        self._weights[signal_name] = (alpha, beta)

    def get_weight(self, signal_name: str) -> float:
        """Thompson sample from Beta posterior.

        Samples a weight from the current posterior — balances exploitation
        (high-precision signals get high weight) with exploration (uncertain
        signals still get occasional high samples, keeping them in play).
        """
        alpha, beta = self._get_posterior(signal_name)
        # random.betavariate is stdlib; no numpy needed
        return random.betavariate(alpha, beta)

    def get_ranking(self) -> list[tuple[str, float, float]]:
        """Return signals ranked by mean effectiveness: (name, mean, confidence).

        Mean = alpha / (alpha + beta).
        Confidence here is a normalised measure of how many observations have
        been made; it saturates toward 1.0 as alpha + beta grows.
        """
        rows: list[tuple[str, float, float]] = []
        for name, (alpha, beta) in self._weights.items():
            n = alpha + beta - 2.0  # subtract the prior counts
            mean = alpha / (alpha + beta)
            obs_confidence = 1.0 - math.exp(-0.1 * max(n, 0.0))
            rows.append((name, round(mean, 4), round(obs_confidence, 4)))
        rows.sort(key=lambda r: r[1], reverse=True)
        return rows

    def save(self, path: str) -> None:
        os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)
        payload = {name: list(ab) for name, ab in self._weights.items()}
        with open(path, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2)

    def load(self, path: str) -> None:
        with open(path, encoding="utf-8") as fh:
            raw: dict[str, list[float]] = json.load(fh)
        self._weights = {name: (ab[0], ab[1]) for name, ab in raw.items()}

File: workers/src/scorer.js
/**
 * scorer.js — Weighted Bayesian scoring engine (JS port of Python scorer)
 *
 * Uses log-odds accumulation matching the Python implementation:
 *   - Starts at log-odds 0 (50% prior)
 *   - Positive signals push toward legitimate (reduce score)
 *   - All other categories push toward scam (increase score)
 *   - Final score = sigmoid(log_odds)
 *
 * Risk thresholds match Python _RISK_THRESHOLDS:
 *   SAFE        < 0.2
 *   LOW         0.2 – 0.4
 *   SUSPICIOUS  0.4 – 0.6
 *   HIGH        0.6 – 0.8
 *   SCAM        >= 0.8
 */

// ---------------------------------------------------------------------------
// Core scoring
// ---------------------------------------------------------------------------

/**
 * Compute scam score and confidence from extracted signals.
 * @param {Array<{name: string, category: string, weight: number, confidence: number}>} signals
 * @returns {{ scam_score: number, confidence: number }}
 */
export function scoreSignals(signals) {
  if (!signals || signals.length === 0) {
    return { scam_score: 0.0, confidence: 0.0 };
  }

  const EPSILON = 1e-6;
  let logOdds = 0.0;

  for (const s of signals) {
    const w = Math.max(EPSILON, Math.min(1.0 - EPSILON, s.weight));

    if (s.category === 'positive') {
      // Positive signals: shift log-odds toward legitimate
      logOdds -= Math.log((1.0 - w) / w);
    } else {
      // Scam signals: shift log-odds toward scam
      logOdds += Math.log(w / (1.0 - w));
    }
  }

  // Sigmoid to get probability
  const scam_score = Math.round((1.0 / (1.0 + Math.exp(-logOdds))) * 10000) / 10000;

  // Confidence: grows with signal count, penalised when signals disagree
  const nScam = signals.filter(s => s.category !== 'positive').length;
  const nPos = signals.filter(s => s.category === 'positive').length;
  const total = signals.length;

  // Base confidence from signal count (asymptote ~0.95)
  const baseConf = 1.0 - Math.exp(-0.3 * total);

  // Agreement factor: 1.0 when all same-direction, lower when mixed
  const majority = Math.max(nScam, nPos);
  const agreement = total > 0 ? majority / total : 1.0;

  const confidence = Math.round(baseConf * agreement * 10000) / 10000;

  return { scam_score, confidence };
}

// ---------------------------------------------------------------------------
// Risk classification
// ---------------------------------------------------------------------------

const RISK_THRESHOLDS = {
  safe: 0.2,
  low: 0.4,
  suspicious: 0.6,
  high: 0.8,
};

/**
 * Map a scam score to a risk level string.
 * @param {number} scamScore
 * @returns {string}
 */
export function classifyRisk(scamScore) {
  if (scamScore < RISK_THRESHOLDS.safe) return 'safe';
  if (scamScore < RISK_THRESHOLDS.low) return 'low';
  if (scamScore < RISK_THRESHOLDS.suspicious) return 'suspicious';
  if (scamScore < RISK_THRESHOLDS.high) return 'high';
  return 'scam';
}

/**
 * Map a scam score to a human-readable risk label.
 * @param {number} scamScore
 * @returns {string}
 */
export function riskLabel(scamScore) {
  if (scamScore < 0.2) return 'Verified Safe';
  if (scamScore < 0.4) return 'Likely Legitimate';
  if (scamScore < 0.6) return 'Suspicious';
  if (scamScore < 0.8) return 'Likely Scam';
  return 'Almost Certainly Scam';
}

// ---------------------------------------------------------------------------
// Full result builder
// ---------------------------------------------------------------------------

/**
 * Build the complete analysis result object (matches Python ValidationResult.to_dict).
 * @param {object} job
 * @param {Array<object>} signals
 * @param {number} analysisTimeMs
 * @returns {object}
 */
export function buildResult(job, signals, analysisTimeMs = 0) {
  const { scam_score, confidence } = scoreSignals(signals);
  const risk_level = classifyRisk(scam_score);
  const risk_label = riskLabel(scam_score);

  // Partition signals by category
  const red_flags = signals
    .filter(s => s.category === 'red_flag')
    .map(s => ({ name: s.name, detail: s.detail, evidence: s.evidence }));

  const warnings = signals
    .filter(s => s.category === 'warning')
    .map(s => ({ name: s.name, detail: s.detail, evidence: s.evidence }));

  const ghost_indicators = signals
    .filter(s => s.category === 'ghost_job')
    .map(s => ({ name: s.name, detail: s.detail, evidence: s.evidence }));

  const positive_signals = signals
    .filter(s => s.category === 'positive')
    .map(s => ({ name: s.name, detail: s.detail, evidence: s.evidence }));

  const structural = signals
    .filter(s => s.category === 'structural')
    .map(s => ({ name: s.name, detail: s.detail, evidence: s.evidence }));

  return {
    job: {
      url: job.url || '',
      title: job.title || '',
      company: job.company || '',
      location: job.location || '',
      salary_min: job.salary_min || 0,
      salary_max: job.salary_max || 0,
      posted_date: job.posted_date || '',
      experience_level: job.experience_level || '',
      employment_type: job.employment_type || '',
      is_remote: Boolean(job.is_remote),
    },
    scam_score: Math.round(scam_score * 1000) / 1000,
    confidence: Math.round(confidence * 1000) / 1000,
    risk_level,
    risk_label,
    red_flags,
    warnings,
    ghost_indicators,
    positive_signals,
    structural,
    signal_count: signals.length,
    ai_tier_used: 'worker-regex',
    analysis_time_ms: Math.round(analysisTimeMs * 10) / 10,
    source: 'cloudflare-worker',
  };
}
